File: driver/python/pysequoiadb/collectionspace.py
#   Copyright (C) 2012-2014 SequoiaDB Ltd.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""Module of collectionspace for python driver of SequoiaDB
"""

try:
    from . import sdb
except:
    raise Exception("Cannot find extension: sdb")

import bson
from bson.py3compat import (str_type)
from pysequoiadb.collection import collection
from pysequoiadb.cursor import cursor
from pysequoiadb.errcode import SDB_OOM
from pysequoiadb.error import (SDBBaseError, SDBSystemError, SDBTypeError, SDBEndOfCursor, raise_if_error)


class collectionspace(object):
    """CollectionSpace for SequoiaDB

    All operation need deal with the error code returned first, if it has.
    Every error code is not SDB_OK(or 0), it means something error has appeared,
    and user should deal with it according the meaning of error code printed.

    @version: execute to get version
              >>> import pysequoiadb
              >>> print pysequoiadb.get_version()

    @notice : The dict of built-in Python is hashed and non-ordered. so the
              element in dict may not the order we make it. we make a dict and
              print it like this:
              ...
              >>> a = {"avg_age":24, "major":"computer science"}
              >>> a
              >>> {'major': 'computer science', 'avg_age': 24}
              ...
              the elements order it is not we make it!!
              therefore, we use bson.SON to make the order-sensitive dict if the
              order is important such as operations in "$sort", "$group",
              "split_by_condition", "aggregate","create_collection"...
              In every scene which the order is important, please make it using
              bson.SON and list. It is a subclass of built-in dict
              and order-sensitive
    """

    def __init__(self):
        """invoked when a new object is producted.

        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        # 'cs' is short for collection space
        try:
            self._cs = sdb.create_cs()
        except SystemError:
            raise SDBSystemError(SDB_OOM, "Failed to alloc collection space")

    def __del__(self):
        """delete a object existed.

        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        if self._cs is not None:
            rc = sdb.release_cs(self._cs)
            raise_if_error(rc, "Failed to release collection space")
            self._cs = None

    def __repr__(self):
        return "Collection Space: %s" % (self.get_collection_space_name())

    def __getattr__(self, name):
        """support client.cs to access to collection.

           eg.
           cc = client()
           cs = cc.test
           cl = cs.test_cl  # access to collection named 'test_cl'

           and we should pass '__members__' and '__methods__',
           becasue dir(cc) will invoke __getattr__("__members__") and
           __getattr__("__methods__").

           if success, a collection object will be returned.

        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        if '__members__' == name or '__methods__' == name:
            pass
        else:
            cl = collection()
            try:
                rc = sdb.cs_get_collection(self._cs, name, cl._cl)
                raise_if_error(rc, "Failed to get collection: %s" % name)
            except SDBBaseError:
                del cl
                raise

            return cl

    def __getitem__(self, name):
        """support [] to access to collection.

           eg.
           cc = client()
           cs = cc['test']
           cl = cs['test_cl']   # access to collection named 'test_cl'.

        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        return self.__getattr__(name)

    def get_collection(self, cl_name):
        """Get the named collection.

        Parameters:
           Name         Type     Info:
           cl_name      str      The short name of the collection.
        Return values:
           a collection object of query
        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        if not isinstance(cl_name, str_type):
            raise SDBTypeError("collection must be an instance of str_type")

        cl = collection()
        try:
            rc = sdb.cs_get_collection(self._cs, cl_name, cl._cl)
            raise_if_error(rc, "Failed to get collection: %s" % cl_name)
        except SDBBaseError:
            del cl
            raise

        return cl

    def create_collection(self, cl_name, options=None):
        """create a collection using name and options.

        Parameters:
           Name      Type     Info:
           cl_name   str      The collection name.
           options   dict     The options are as following:
                              ShardingKey          : Assign the sharding key, foramt: { ShardingKey: { <key name>: <1/-1>} },
                                                     1 indicates positive order, -1 indicates reverse order.
                                                     e.g. { ShardingKey: { age: 1 } }
                              ShardingType         : Assign the sharding type, default is "hash"
                              Partition            : The number of partition, it is valid when ShardingType is "hash",
                                                     the range is [2^3, 2^20], default is 4096
                              ReplSize             : Assign how many replica nodes need to be synchronized when a write
                                                     request (insert, update, etc) is executed, default is 1
                              Compressed           : Whether to enable data compression, default is true
                              CompressionType      : The compression type of data, could be "snappy" or "lzw", default is "lzw"
                              AutoSplit            : Whether to enable the automatic partitioning, it is valid when ShardingType
                                                     is "hash", defalut is false
                              Group                : Assign the data group to which it belongs, default: The collection will
                                                     be created in any data group of the domain that the collection belongs to
                              AutoIndexId          : Whether to build "$id" index, default is true
                              EnsureShardingIndex  : Whether to build sharding index, default is true
                              StrictDataMode       : Whether to enable strict date mode in numeric operations, default is false
                              AutoIncrement        : Assign attributes of an autoincrement field or batch autoincrement fields
                                                     e.g. { AutoIncrement : { Field : "a", MaxValue : 2000 } },
                                                     { AutoIncrement : [ { Field : "a", MaxValue : 2000}, { Field : "a", MaxValue : 4000 } ] }
                              LobShardingKeyFormat : Assign the format of lob sharding key, could be "YYYYMMDD", "YYYYMM" or "YYYY".
                                                     It is valid when the collection is main collection
                              IsMainCL             : Main collection or not, default is false, which means it is not main collection
                              DataSource           : The name of the date soure used
                              Mapping              : The name of the collection to be mapped
        Return values:
           a collection object created
        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        if not isinstance(cl_name, str_type):
            raise SDBTypeError("collection must be an instance of str_type")

        bson_options = None
        if options is not None:
            if not isinstance(options, dict):
                raise SDBTypeError("options must be an instance of dict")
            bson_options = bson.BSON.encode(options)

        cl = collection()
        try:
            if bson_options is None:
                rc = sdb.cs_create_collection(self._cs, cl_name, cl._cl)
            else:
                rc = sdb.cs_create_collection_use_opt(self._cs, cl_name,
                                                      bson_options, cl._cl)
            raise_if_error(rc, "Failed to create collection")
        except SDBBaseError:
            del cl
            raise

        return cl

    def drop_collection(self, cl_name):
        """Drop the specified collection in current collection space.

        Parameters:
           Name      Type     Info:
           cl_name   str      The collection name.
        Exceptions:
           pysequoiadb.error.SDBTypeError
           pysequoiadb.error.SDBBaseError
        """
        if not isinstance(cl_name, str_type):
            raise SDBTypeError("collection must be an instance of str_type")

        rc = sdb.cs_drop_collection(self._cs, cl_name)
        raise_if_error(rc, "Failed to drop collection")

    def rename_collection(self, old_name, new_name, options=None):
        """Rename the specified collection in current collection space.

        Parameters:
           Name      Type     Info:
           old_name  str      The original name of collection.
           new_name  str      The new name of collection.
           options   dict     Options for renaming.
        Exceptions:
           pysequoiadb.error.SDBTypeError
           pysequoiadb.error.SDBBaseError
        """
        if not isinstance(old_name, str_type):
            raise SDBTypeError("old name must be an instance of str_type")
        if not isinstance(new_name, str_type):
            raise SDBTypeError("new name must be an instance of str_type")
        bson_options = None
        if options is not None:
            if not isinstance(options, dict):
                raise SDBTypeError("options must be an instance of dict")
            bson_options = bson.BSON.encode(options)

        rc = sdb.cs_rename_collection(self._cs, old_name, new_name, bson_options)
        raise_if_error(rc, "Failed to rename collection")

    def get_collection_names(self):
        """Get the names of all collections in the current collection space.

        Return values:
           A list of collection names
        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        result = []
        tmp = cursor()

        try:
            rc = sdb.cs_get_collection_names(self._cs, tmp._cursor)
            raise_if_error(rc, "Failed to all collection in the current collection space")
        except SDBBaseError:
            del tmp
            raise

        try:
            while True:
                try:
                    record = tmp.next()
                    result.append(record['Name'])
                except SDBEndOfCursor:
                    break
        finally:
            tmp.close()
            del tmp

        return result

    def get_collection_space_name(self):
        """Get the current collection space name.

        Return values:
           The name of current collection space.
        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        rc, cs_name = sdb.cs_get_collection_space_name(self._cs)
        raise_if_error(rc, "Failed to get collection space name")

        return cs_name

    def alter(self, options):
        """Alter the current collection space.

        Parameters:
           Name     Type           Info:
           options   dict          The options for alter collection space, including
                                   Domain      : domain of collection space
                                   PageSize    : page size of collection space
                                   LobPageSize : LOB page size of collection space
        """
        if not isinstance(options, dict):
            raise SDBTypeError("options must be an instance of dict")
        bson_options = bson.BSON.encode(options)

        rc = sdb.cs_alter(self._cs, bson_options)
        raise_if_error(rc, "Failed to alter collection space")

    def set_domain(self, options):
        """Alter the current collection space to set domain.

        Parameters:
           Name     Type           Info:
           options   dict          The options for alter collection space, including
                                 Domain      : domain of collection space
        """
        if not isinstance(options, dict):
            raise SDBTypeError("options must be an instance of dict")
        bson_options = bson.BSON.encode(options)

        rc = sdb.cs_set_domain(self._cs, bson_options)
        raise_if_error(rc, "Failed to alter collection space to set domain")

    def remove_domain(self):
        """Alter the current collection space to remove domain.
        """
        rc = sdb.cs_remove_domain(self._cs)
        raise_if_error(rc, "Failed to alter collection space to remove domain")

    def get_domain_name(self):
        """Get the Domain name of the current collection space. Returns an empty string
           if the current collection space

        Return values:
           The domain name of the current collection space.
        Exceptions:
           pysequoiadb.error.SDBBaseError
        """
        rc, domain_name = sdb.cs_get_domain_name(self._cs)
        raise_if_error(rc, "Failed to get domain name of the current collection space")

        return domain_name

    def enable_capped(self):
        """Alter the current collection space to enable capped.
        """
        rc = sdb.cs_enable_capped(self._cs)
        raise_if_error(rc, "Failed to alter collection space to enable capped")

    def disable_capped(self):
        """Alter the current collection space to disble capped.
        """
        rc = sdb.cs_disable_capped(self._cs)
        raise_if_error(rc, "Failed to alter collection space to disable capped")

    def set_attributes(self, options):
        """Alter the current collection space.

        Parameters:
           Name     Type           Info:
           options   dict          The options for alter collection space, including
                                   Domain      : domain of collection space
                                   PageSize    : page size of collection space
                                   LobPageSize : LOB page size of collection space
        """
        if not isinstance(options, dict):
            raise SDBTypeError("options must be an instance of dict")
        bson_options = bson.BSON.encode(options)

        rc = sdb.cs_set_attributes(self._cs, bson_options)
        raise_if_error(rc, "Failed to alter collection space")

File: driver/java/src/main/java/com/sequoiadb/base/CollectionSpace.java
/*
 * Copyright 2018 SequoiaDB Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.sequoiadb.base;

import com.sequoiadb.exception.BaseException;
import com.sequoiadb.exception.SDBError;
import com.sequoiadb.message.request.AdminRequest;
import com.sequoiadb.message.response.SdbReply;
import org.bson.BSONObject;
import org.bson.BasicBSONObject;
import org.bson.types.BasicBSONList;

import java.util.ArrayList;
import java.util.List;

/**
 * Collection space of SequoiaDB.
 */
public class CollectionSpace {
    private String name;
    private Sequoiadb sequoiadb;

    private static final String FIELD_NAME_DOMAIN = "Domain";

    /**
     * Return the name of current collection space.
     *
     * @return The collection space name
     */
    public String getName() {
        return name;
    }

    /**
     * Get the Sequoiadb instance of current collection space belongs to.
     *
     * @return Sequoiadb object
     */
    public Sequoiadb getSequoiadb() {
        return sequoiadb;
    }

    /**
     * @param sequoiadb Sequoiadb instance
     * @param name      Collection space name
     */
    CollectionSpace(Sequoiadb sequoiadb, String name) {
        this.name = name;
        this.sequoiadb = sequoiadb;
    }

    /**
     * Get the named collection.
     *
     * @param collectionName The collection name
     * @return the object of the specified collection, or an exception when the collection does not exist.
     * @throws BaseException If error happens.
     */
    public DBCollection getCollection(String collectionName) throws BaseException {
        if (collectionName == null || collectionName.equals("")) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "collectionName can't be null or empty");
        }
        // get cl from cache
        String collectionFullName = name + "." + collectionName;
        if (sequoiadb.fetchCache(collectionFullName)) {
            return new DBCollection(sequoiadb, this, collectionName);
        }
        // create cl and then update cache
        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);
        AdminRequest request = new AdminRequest(AdminCommand.TEST_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);
        sequoiadb.throwIfError(response);
        sequoiadb.upsertCache(collectionFullName);
        return new DBCollection(sequoiadb, this, collectionName);
    }

    /**
     * Verify the existence of collection in current collection space.
     *
     * @param collectionName The collection name
     * @return True if collection existed or False if not existed
     * @throws BaseException If error happens.
     */
    public boolean isCollectionExist(String collectionName) throws BaseException {
        if (collectionName == null || collectionName.equals("")) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "collectionName can't be null or empty");
        }
        String collectionFullName = name + "." + collectionName;

        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);

        AdminRequest request = new AdminRequest(AdminCommand.TEST_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);

        int flag = response.getFlag();
        if (flag == 0) {
            sequoiadb.upsertCache(collectionFullName);
            return true;
        } else if (flag == SDBError.SDB_DMS_NOTEXIST.getErrorCode()) {
            sequoiadb.removeCache(collectionFullName);
            return false;
        } else {
            sequoiadb.throwIfError(response);
            return false; // make compiler happy
        }
    }

    /**
     * Get all the collection names of current collection space.
     *
     * @return A list of collection names
     * @throws BaseException If error happens.
     */
    public List<String> getCollectionNames() throws BaseException {
        ArrayList<String> colList = new ArrayList<>();

        BSONObject condition = new BasicBSONObject();
        BSONObject subObj = new BasicBSONObject();

        subObj.put("$gt", name + ".");
        subObj.put("$lt", name + "/");

        condition.put(SdbConstants.FIELD_NAME_NAME, subObj);

        try(DBCursor cursor = sequoiadb.getList(Sequoiadb.SDB_LIST_COLLECTIONS, condition, null, null)) {
            while (cursor.hasNext()) {
                colList.add(cursor.getNext().get("Name").toString());
            }
        }
        return colList;
    }

    /**
     * Create the named collection in current collection space.
     *
     * @param collectionName The collection name
     * @return the newly created object of collection
     * @throws BaseException If error happens.
     */
    public DBCollection createCollection(String collectionName)
            throws BaseException {
        return createCollection(collectionName, null);
    }

    /**
     * Create collection by options.
     *
     * @param collectionName The collection name
     * @param options The options are as following:
     *               <ul>
     *               <li>ShardingKey : Assign the sharding key, foramt: { ShardingKey: { <key name>: <1/-1>} }, 
     *               1 indicates positive order, -1 indicates reverse order. e.g. { ShardingKey: { age: 1 } }
     *               <li>ShardingType : Assign the sharding type, default is "hash"
     *               <li>Partition : The number of partition, it is valid when ShardingType is "hash",
     *               the range is [2^3, 2^20], default is 4096
     *               <li>ReplSize : Assign how many replica nodes need to be synchronized when a write
     *               request (insert, update, etc) is executed, default is 1
     *               <li>Compressed : Whether to enable data compression, default is true
     *               <li>CompressionType : The compression type of data, could be "snappy" or "lzw", default is "lzw"
     *               <li>AutoSplit : Whether to enable the automatic partitioning, it is valid when ShardingType
     *               is "hash", defalut is false
     *               <li>Group: Assign the data group to which it belongs, default: The collection will
     *               be created in any data group of the domain that the collection belongs to
     *               <li>AutoIndexId : Whether to build "$id" index, default is true
     *               <li>EnsureShardingIndex : Whether to build sharding index, default is true
     *               <li>StrictDataMode : Whether to enable strict date mode in numeric operations, default is false
     *               <li>AutoIncrement : Assign attributes of an autoincrement field or batch autoincrement fields
     *               e.g. { AutoIncrement : { Field : "a", MaxValue : 2000 } },
     *               { AutoIncrement : [ { Field : "a", MaxValue : 2000}, { Field : "a", MaxValue : 4000 } ] }
     *               <li>LobShardingKeyFormat : Assign the format of lob sharding key, could be "YYYYMMDD", "YYYYMM" or "YYYY".
     *               It is valid when the collection is main collection
     *               <li>IsMainCL : Main collection or not, default is false, which means it is not main collection
     *               <li>DataSource : The name of the date soure used
     *               <li>Mapping : The name of the collection to be mapped
     * @return the newly created object of collection.
     * @throws BaseException If error happens.
     */
    public DBCollection createCollection(String collectionName, BSONObject options) {
        if (collectionName == null || collectionName.equals("")) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "collectionName can't be null or empty");
        }
        String collectionFullName = name + "." + collectionName;
        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);
        if (options != null) {
            obj.putAll(options);
        }

        AdminRequest request = new AdminRequest(AdminCommand.CREATE_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);
        String msg = "collection = " + collectionFullName + ", options = " + options;
        sequoiadb.throwIfError(response, msg);
        sequoiadb.upsertCache(collectionFullName);
        return new DBCollection(sequoiadb, this, collectionName);
    }

    /**
     * Drop current collection space.
     *
     * @throws BaseException If error happens.
     * @see com.sequoiadb.base.Sequoiadb#dropCollectionSpace(String)
     * @deprecated Use Sequoiadb.dropCollectionSpace() instead.
     */
    public void drop() throws BaseException {
        sequoiadb.dropCollectionSpace(this.name);
    }

    /**
     * Remove the named collection of current collection space.
     *
     * @param collectionName The collection name
     * @throws BaseException If error happens.
     */
    public void dropCollection(String collectionName) throws BaseException {
        dropCollection(collectionName, null);
    }

    /**
     * Remove the named collection of current collection space.
     *
     * @param collectionName The collection name
     * @param options The options for drop specified collection
     *                <ul>
     *                <li>SkipRecycleBin : Indicates whether to skip recycle bin, default is false.
     *                </ul>
     * @throws BaseException If error happens.
     */
    public void dropCollection(String collectionName, BSONObject options)
            throws BaseException {
        if (collectionName == null || collectionName.equals("")) {
            throw new BaseException(SDBError.SDB_INVALIDARG,
                                    "collectionName can't be null or empty");
        }
        String collectionFullName = name + "." + collectionName;
        BSONObject rebuildOptions = new BasicBSONObject();
        rebuildOptions.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);
        if (options != null) {
            rebuildOptions.putAll(options);
        }
        AdminRequest request = new AdminRequest(AdminCommand.DROP_CL,
                                                rebuildOptions);
        SdbReply response = sequoiadb.requestAndResponse(request);
        sequoiadb.throwIfError(response, collectionName);
        sequoiadb.removeCache(collectionFullName);
    }

    private void alterInternal(String taskName, BSONObject options, boolean allowNullArgs) throws BaseException {
        if (null == options && !allowNullArgs) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "options is null");
        }
        BSONObject argumentObj = new BasicBSONObject();
        argumentObj.put(SdbConstants.FIELD_NAME_NAME, taskName);
        argumentObj.put(SdbConstants.FIELD_NAME_ARGS, options);
        BSONObject alterObject = new BasicBSONObject();
        alterObject.put(SdbConstants.FIELD_NAME_ALTER, argumentObj);
        alterCollectionSpace(alterObject);
    }

    /**
     * Alter the current collection space.
     *
     * @param options The options of the collection space:
     *                <ul>
     *                <li>Domain : the domain of collection space.
     *                <li>PageSize : page size of collection space.
     *                <li>LobPageSize : LOB page size of collection space.
     *                </ul>
     * @throws BaseException If error happens.
     */
    public void alterCollectionSpace(BSONObject options) throws BaseException {
        if (null == options) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "options is null");
        }

        BSONObject newObj = new BasicBSONObject();
        if (!options.containsField(SdbConstants.FIELD_NAME_ALTER)) {
            newObj.put(SdbConstants.FIELD_NAME_NAME, name);
            newObj.put(SdbConstants.FIELD_NAME_OPTIONS, options);
        } else {
            Object tmpAlter = options.get(SdbConstants.FIELD_NAME_ALTER);
            if (tmpAlter instanceof BasicBSONObject ||
                    tmpAlter instanceof BasicBSONList) {
                newObj.put(SdbConstants.FIELD_NAME_ALTER, tmpAlter);
            } else {
                throw new BaseException(SDBError.SDB_INVALIDARG, options.toString());
            }
            newObj.put(SdbConstants.FIELD_NAME_ALTER_TYPE, SdbConstants.SDB_ALTER_CS);
            newObj.put(SdbConstants.FIELD_NAME_VERSION, SdbConstants.SDB_ALTER_VERSION);
            newObj.put(SdbConstants.FIELD_NAME_NAME, name);

            if (options.containsField(SdbConstants.FIELD_NAME_OPTIONS)) {
                Object tmpOptions = options.get(SdbConstants.FIELD_NAME_OPTIONS);
                if (tmpOptions instanceof BasicBSONObject) {
                    newObj.put(SdbConstants.FIELD_NAME_OPTIONS, tmpOptions);
                } else {
                    throw new BaseException(SDBError.SDB_INVALIDARG, options.toString());
                }
            }
        }

        AdminRequest request = new AdminRequest(AdminCommand.ALTER_CS, newObj);
        SdbReply response = sequoiadb.requestAndResponse(request);
        sequoiadb.throwIfError(response);
    }

    /**
     * Alter the current collection space to set domain
     *
     * @param options The options of the collection space:
     *                <ul>
     *                <li>Domain : the domain of collection space.
     *                </ul>
     * @throws BaseException If error happens.
     */
    public void setDomain(BSONObject options) throws BaseException {
        alterInternal(SdbConstants.SDB_ALTER_SET_DOMAIN, options, false);
    }

    /**
     * Alter the current collection space to remove domain
     *
     * @throws BaseException If error happens.
     */
    public void removeDomain() throws BaseException {
        alterInternal(SdbConstants.SDB_ALTER_REMOVE_DOMAIN, null, true);
    }

    /**
     * Get the Domain name of the current collection space. Returns an empty string if the current collection space
     * has no owning domain
     * @return the Domain name.
     * @throws BaseException If error happens.
     */
    public String getDomainName(){
        String result = null;
        String cmd = "select Domain from $LIST_CS where Name = '" + this.name + "'";
        DBCursor cursor = sequoiadb.exec(cmd);
        try
        {
            if (cursor.hasNext()) {
                BSONObject obj = cursor.getNext();
                result = (String) obj.get(FIELD_NAME_DOMAIN);
            }
            return result == null ? "" : result;
        }
        finally
        {
            cursor.close();
        }
    }

    /**
     * Alter the current collection space to enable capped
     *
     * @throws BaseException If error happens.
     */
    public void enableCapped() throws BaseException {
        alterInternal(SdbConstants.SDB_ALTER_ENABLE_CAPPED, null, true);
    }

    /**
     * Alter the current collection space to disable capped
     *
     * @throws BaseException If error happens.
     */
    public void disableCapped() throws BaseException {
        alterInternal(SdbConstants.SDB_ALTER_DISABLE_CAPPED, null, true);
    }

    /**
     * Alter the current collection space to set attributes.
     *
     * @param options The options of the collection space:
     *                <ul>
     *                <li>Domain : the domain of collection space.
     *                <li>PageSize : page size of collection space.
     *                <li>LobPageSize : LOB page size of collection space.
     *                </ul>
     * @throws BaseException If error happens.
     */
    public void setAttributes(BSONObject options) throws BaseException {
        alterInternal(SdbConstants.SDB_ALTER_SET_ATTRIBUTES, options, false);
    }

    /**
     * @param oldName The old collection name
     * @param newName The new collection name
     * @throws BaseException If error happens.
     */
    public void renameCollection(String oldName, String newName) throws BaseException {
        if (oldName == null || oldName.length() == 0) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "The old name of collection is null or empty");
        }
        if (newName == null || newName.length() == 0) {
            throw new BaseException(SDBError.SDB_INVALIDARG, "The new name of collection is null or empty");
        }

        BSONObject matcher = new BasicBSONObject();
        matcher.put(SdbConstants.FIELD_NAME_CELLECTIONSPACE, name);
        matcher.put(SdbConstants.FIELD_NAME_OLDNAME, oldName);
        matcher.put(SdbConstants.FIELD_NAME_NEWNAME, newName);

        AdminRequest request = new AdminRequest(AdminCommand.RENAME_CL, matcher);
        SdbReply response = sequoiadb.requestAndResponse(request);
        sequoiadb.throwIfError(response);
        sequoiadb.removeCache(name + "." + oldName);
        sequoiadb.upsertCache(name + "." + newName);
    }
}
